compiler/rustc_session/src/options.rs

```diff
@@ -203,6 +204,9 @@ top_level_options!(
         json_unused_externs: bool [UNTRACKED],
 
         pretty: Option<PpMode> [UNTRACKED],
+
+        /// The (potentially remapped) working directory
+        working_dir: RealFileName [TRACKED],
```

Comment:
This will result in a different hash for `/wd/foo.rs` vs `/baz/foo.rs` with `/baz` remapped to `/wd` despite the filename after any remapping being equal.

Reply:
I don't think so - this field is initiwlized using the remapped working directory.

Reply:
`RealFileName` is an enum defined as

```rust
#[derive(HashStable_Generic)]
enum RealFileName {
    LocalPath(PathBuf),
    Remapped {
        local_path: Option<PathBuf>,
        virtual_name: PathBuf,
    },
}
```

while the `Hash` implementation only looks at the remapped path, the `HashStable` implementation is derived and looks at which enum variant it is and in case of a remapped path also at the local path before remapping. The dep tracking hash is calculated using `HashStable` and not `Hash`, so this will depend if the path is remapped or not and more importantly the local path before remapping, which is even worse that I thought as it breaks reproducible builds.

Reply:
> The dep tracking hash is calculated using HashStable and not Hash

That's incorrect - I'm using `impl_dep_tracking_hash_via_hash` for `RealFileName`: https://github.com/Aaron1011/rust/blob/47daf829b1a0d2d499745a34cf4b1969929e5f7f/compiler/rustc_session/src/config.rs#L2513

It looks like the `HashStable` implementations for `FileName` and `RealFileName` are unused, and can be removed. Having `Hash, `HashStable,` and `DepTrackingHash` all implemented for a single type is pretty confusing - I wonder if there's any kind of compile-time check we can do to prevent that.